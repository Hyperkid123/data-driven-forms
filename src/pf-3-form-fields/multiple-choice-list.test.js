import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Form, Field } from 'react-final-form';
import ComponentType, { configureContext } from '../renderer-context';
import MultipleChoiceList from './multiple-choice-list';

describe('Pf3 <MultipleChoiceList />', () => {
  let FormWrapper;
  let multipleChoiceDefinition;
  beforeEach(() => {
    FormWrapper = ({ initialValues, children }) => ( // eslint-disable-line
      <Form onSubmit={ jest.fn() } initialValues={ initialValues }>
        { () => (
          <ComponentType.Provider value={ configureContext({ formType: 'pf3' }) }>
            { children }
          </ComponentType.Provider>
        ) }
      </Form>
    );
    multipleChoiceDefinition = {
      key: 'multipleChoicesList',
      title: 'A multiple choices list',
      name: 'multipleChoicesList',
      label: 'A multiple choices list',
      component: 'checkbox-field',
      dataType: 'string',
      validate: [],
      options: [{
        label: 'foo',
        value: 'foo',
      }, {
        value: 'bar',
        label: 'bar',
      }, {
        value: 'fuzz',
        label: 'fuzz',
      }, {
        value: 'qux',
        label: 'qux',
      },
      ],
      type: 'checkbox',
    };
  });

  it('should render correctly', () => {
    const wrapper = mount(
      <FormWrapper>
        <MultipleChoiceList { ...multipleChoiceDefinition }/>
      </FormWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly with error message', () => {
    const wrapper = mount(
      <FormWrapper>
        <MultipleChoiceList { ...multipleChoiceDefinition } validate={ [ () => 'Error message' ] }/>
      </FormWrapper>
    );
    wrapper.find(Field).first().instance().setState(prevState => ({
      state: {
        ...prevState.state,
        error: 'Has some error',
        touched: true,
      },
    }));
    wrapper.find(Form).first().update();
    expect(toJson(wrapper.find(Field).first())).toMatchSnapshot();
  });

  it('should correctly add value to formState', () => {
    const wrapper = mount(
      <FormWrapper>
        <MultipleChoiceList { ...multipleChoiceDefinition }/>
      </FormWrapper>
    );

    // simulate click on checkbox with value qux
    wrapper.find('input[type="checkbox"]').last().simulate('change', { target: { checked: true }});
    expect(wrapper.find(Form).first().instance().state.state.values.multipleChoicesList).toEqual([ 'qux' ]);
  });

  it('should correctly remove value from formState', () => {
    const wrapper = mount(
      <FormWrapper initialValues={{
        multipleChoicesList: [ 'foo', 'bar', 'fuzz', 'qux' ],
      }}>
        <MultipleChoiceList { ...multipleChoiceDefinition }/>
      </FormWrapper>
    );

    // simulate click on checkbox with value fuzz
    wrapper.find('input[type="checkbox"]').at(2).simulate('change', { target: { checked: false }});
    expect(wrapper.find(Form).first().instance().state.state.values.multipleChoicesList).toEqual([ 'foo', 'bar', 'qux' ]);
  });
});
