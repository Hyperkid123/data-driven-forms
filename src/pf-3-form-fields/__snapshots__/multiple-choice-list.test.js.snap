// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pf3 <MultipleChoiceList /> should render correctly 1`] = `
<FormWrapper>
  <ReactFinalForm
    onSubmit={[MockFunction]}
  >
    <MultipleChoiceList
      component="checkbox-field"
      dataType="string"
      key="multipleChoicesList"
      label="A multiple choices list"
      name="multipleChoicesList"
      options={
        Array [
          Object {
            "label": "foo",
            "value": "foo",
          },
          Object {
            "label": "bar",
            "value": "bar",
          },
          Object {
            "label": "fuzz",
            "value": "fuzz",
          },
          Object {
            "label": "qux",
            "value": "qux",
          },
        ]
      }
      title="A multiple choices list"
      type="checkbox"
      validate={Array []}
    >
      <Field
        component="checkbox-field"
        dataType="string"
        format={[Function]}
        label="A multiple choices list"
        name="multipleChoicesList"
        options={
          Array [
            Object {
              "label": "foo",
              "value": "foo",
            },
            Object {
              "label": "bar",
              "value": "bar",
            },
            Object {
              "label": "fuzz",
              "value": "fuzz",
            },
            Object {
              "label": "qux",
              "value": "qux",
            },
          ]
        }
        parse={[Function]}
        title="A multiple choices list"
        type="checkbox"
        validate={[Function]}
      >
        <FormGroup
          bsClass="form-group"
          validationState={null}
        >
          <div
            className="form-group"
          >
            <Col
              bsClass="col"
              className="control-label"
              componentClass="label"
              md={2}
            >
              <label
                className="control-label col-md-2"
              >
                A multiple choices list
              </label>
            </Col>
            <Col
              bsClass="col"
              componentClass="div"
              md={10}
            >
              <div
                className="col-md-10"
              >
                <Field
                  format={[Function]}
                  id="undefined-foo"
                  key="foo"
                  label="foo"
                  name="multipleChoicesList"
                  parse={[Function]}
                  render={[Function]}
                  type="checkbox"
                  value="foo"
                >
                  <Checkbox
                    aria-label="foo"
                    bsClass="checkbox"
                    checked={false}
                    disabled={false}
                    id="undefined-foo"
                    inline={false}
                    label="foo"
                    name="multipleChoicesList"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    title=""
                    type="checkbox"
                    value="foo"
                  >
                    <div
                      className="checkbox"
                    >
                      <label
                        title=""
                      >
                        <input
                          aria-label="foo"
                          checked={false}
                          disabled={false}
                          id="undefined-foo"
                          label="foo"
                          name="multipleChoicesList"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          type="checkbox"
                          value="foo"
                        />
                        foo
                      </label>
                    </div>
                  </Checkbox>
                </Field>
                <Field
                  format={[Function]}
                  id="undefined-bar"
                  key="bar"
                  label="bar"
                  name="multipleChoicesList"
                  parse={[Function]}
                  render={[Function]}
                  type="checkbox"
                  value="bar"
                >
                  <Checkbox
                    aria-label="bar"
                    bsClass="checkbox"
                    checked={false}
                    disabled={false}
                    id="undefined-bar"
                    inline={false}
                    label="bar"
                    name="multipleChoicesList"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    title=""
                    type="checkbox"
                    value="bar"
                  >
                    <div
                      className="checkbox"
                    >
                      <label
                        title=""
                      >
                        <input
                          aria-label="bar"
                          checked={false}
                          disabled={false}
                          id="undefined-bar"
                          label="bar"
                          name="multipleChoicesList"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          type="checkbox"
                          value="bar"
                        />
                        bar
                      </label>
                    </div>
                  </Checkbox>
                </Field>
                <Field
                  format={[Function]}
                  id="undefined-fuzz"
                  key="fuzz"
                  label="fuzz"
                  name="multipleChoicesList"
                  parse={[Function]}
                  render={[Function]}
                  type="checkbox"
                  value="fuzz"
                >
                  <Checkbox
                    aria-label="fuzz"
                    bsClass="checkbox"
                    checked={false}
                    disabled={false}
                    id="undefined-fuzz"
                    inline={false}
                    label="fuzz"
                    name="multipleChoicesList"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    title=""
                    type="checkbox"
                    value="fuzz"
                  >
                    <div
                      className="checkbox"
                    >
                      <label
                        title=""
                      >
                        <input
                          aria-label="fuzz"
                          checked={false}
                          disabled={false}
                          id="undefined-fuzz"
                          label="fuzz"
                          name="multipleChoicesList"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          type="checkbox"
                          value="fuzz"
                        />
                        fuzz
                      </label>
                    </div>
                  </Checkbox>
                </Field>
                <Field
                  format={[Function]}
                  id="undefined-qux"
                  key="qux"
                  label="qux"
                  name="multipleChoicesList"
                  parse={[Function]}
                  render={[Function]}
                  type="checkbox"
                  value="qux"
                >
                  <Checkbox
                    aria-label="qux"
                    bsClass="checkbox"
                    checked={false}
                    disabled={false}
                    id="undefined-qux"
                    inline={false}
                    label="qux"
                    name="multipleChoicesList"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    title=""
                    type="checkbox"
                    value="qux"
                  >
                    <div
                      className="checkbox"
                    >
                      <label
                        title=""
                      >
                        <input
                          aria-label="qux"
                          checked={false}
                          disabled={false}
                          id="undefined-qux"
                          label="qux"
                          name="multipleChoicesList"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          type="checkbox"
                          value="qux"
                        />
                        qux
                      </label>
                    </div>
                  </Checkbox>
                </Field>
              </div>
            </Col>
          </div>
        </FormGroup>
      </Field>
    </MultipleChoiceList>
  </ReactFinalForm>
</FormWrapper>
`;

exports[`Pf3 <MultipleChoiceList /> should render correctly with error message 1`] = `
<Field
  component="checkbox-field"
  dataType="string"
  format={[Function]}
  label="A multiple choices list"
  name="multipleChoicesList"
  options={
    Array [
      Object {
        "label": "foo",
        "value": "foo",
      },
      Object {
        "label": "bar",
        "value": "bar",
      },
      Object {
        "label": "fuzz",
        "value": "fuzz",
      },
      Object {
        "label": "qux",
        "value": "qux",
      },
    ]
  }
  parse={[Function]}
  title="A multiple choices list"
  type="checkbox"
  validate={[Function]}
>
  <FormGroup
    bsClass="form-group"
    validationState="error"
  >
    <div
      className="form-group has-error"
    >
      <Col
        bsClass="col"
        className="control-label"
        componentClass="label"
        md={2}
      >
        <label
          className="control-label col-md-2"
        >
          A multiple choices list
        </label>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        md={10}
      >
        <div
          className="col-md-10"
        >
          <Field
            format={[Function]}
            id="undefined-foo"
            key="foo"
            label="foo"
            name="multipleChoicesList"
            parse={[Function]}
            render={[Function]}
            type="checkbox"
            value="foo"
          >
            <Checkbox
              aria-label="foo"
              bsClass="checkbox"
              checked={false}
              disabled={false}
              id="undefined-foo"
              inline={false}
              label="foo"
              name="multipleChoicesList"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              title=""
              type="checkbox"
              value="foo"
            >
              <div
                className="checkbox"
              >
                <label
                  title=""
                >
                  <input
                    aria-label="foo"
                    checked={false}
                    disabled={false}
                    id="undefined-foo"
                    label="foo"
                    name="multipleChoicesList"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="checkbox"
                    value="foo"
                  />
                  foo
                </label>
              </div>
            </Checkbox>
          </Field>
          <Field
            format={[Function]}
            id="undefined-bar"
            key="bar"
            label="bar"
            name="multipleChoicesList"
            parse={[Function]}
            render={[Function]}
            type="checkbox"
            value="bar"
          >
            <Checkbox
              aria-label="bar"
              bsClass="checkbox"
              checked={false}
              disabled={false}
              id="undefined-bar"
              inline={false}
              label="bar"
              name="multipleChoicesList"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              title=""
              type="checkbox"
              value="bar"
            >
              <div
                className="checkbox"
              >
                <label
                  title=""
                >
                  <input
                    aria-label="bar"
                    checked={false}
                    disabled={false}
                    id="undefined-bar"
                    label="bar"
                    name="multipleChoicesList"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="checkbox"
                    value="bar"
                  />
                  bar
                </label>
              </div>
            </Checkbox>
          </Field>
          <Field
            format={[Function]}
            id="undefined-fuzz"
            key="fuzz"
            label="fuzz"
            name="multipleChoicesList"
            parse={[Function]}
            render={[Function]}
            type="checkbox"
            value="fuzz"
          >
            <Checkbox
              aria-label="fuzz"
              bsClass="checkbox"
              checked={false}
              disabled={false}
              id="undefined-fuzz"
              inline={false}
              label="fuzz"
              name="multipleChoicesList"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              title=""
              type="checkbox"
              value="fuzz"
            >
              <div
                className="checkbox"
              >
                <label
                  title=""
                >
                  <input
                    aria-label="fuzz"
                    checked={false}
                    disabled={false}
                    id="undefined-fuzz"
                    label="fuzz"
                    name="multipleChoicesList"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="checkbox"
                    value="fuzz"
                  />
                  fuzz
                </label>
              </div>
            </Checkbox>
          </Field>
          <Field
            format={[Function]}
            id="undefined-qux"
            key="qux"
            label="qux"
            name="multipleChoicesList"
            parse={[Function]}
            render={[Function]}
            type="checkbox"
            value="qux"
          >
            <Checkbox
              aria-label="qux"
              bsClass="checkbox"
              checked={false}
              disabled={false}
              id="undefined-qux"
              inline={false}
              label="qux"
              name="multipleChoicesList"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              title=""
              type="checkbox"
              value="qux"
            >
              <div
                className="checkbox"
              >
                <label
                  title=""
                >
                  <input
                    aria-label="qux"
                    checked={false}
                    disabled={false}
                    id="undefined-qux"
                    label="qux"
                    name="multipleChoicesList"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="checkbox"
                    value="qux"
                  />
                  qux
                </label>
              </div>
            </Checkbox>
          </Field>
        </div>
      </Col>
    </div>
  </FormGroup>
</Field>
`;
